### PR TITLE
Reinforce protocol related unit tests

### PR DESCRIPTION
**Pull request by limebell**
Added `RoutingTableTest.Generate()`, `RoutingTableTest.PeersToBroadcast()`, `RoutingTableTest.PeersToRefresh()` and `ProtocolTest.RefreshTable()`.

### CODE REVIEW

File: Libplanet.Tests/TestUtils.cs [hunk 1]
```diff
@@ -395,5 +396,20 @@ public static string ToString(BitArray bitArray)
...
         }
+
+        public static PrivateKey GeneratePrivateKeyOfBucketIndex(Address tableAddress, int target)
+        {
+            var table = new RoutingTable(tableAddress);
+            var count = 0;
+            PrivateKey privateKey;
+            do
+            {
+                count++;
```
Comment (dahlia) [inline Libplanet.Tests/TestUtils.cs:407]: What is this counter used for?

File: Libplanet/Net/Protocols/RoutingTable.cs [hunk 2]
```diff
@@ -246,6 +232,22 @@ public void Check(BoundPeer peer, DateTimeOffset start, DateTimeOffset end)
...
+        internal void AddPeer(BoundPeer peer, DateTimeOffset updated)
+        {
+            if (peer is null)
+            {
+                throw new ArgumentNullException(nameof(peer));
+            }
+
+            if (peer.Address.Equals(_address))
+            {
+                throw new ArgumentException("Cannot add self to routing table.");
```
Comment (dahlia) [inline Libplanet/Net/Protocols/RoutingTable.cs:244]: This piece of code is not recently written, but…

```suggestion
                throw new ArgumentException(
                    "A peer is disallowed to add itself to its routing table.",
                    nameof(peer)
                );
```

File: Libplanet/Net/Protocols/RoutingTable.cs [hunk 3]
```diff
@@ -246,6 +232,22 @@ public void Check(BoundPeer peer, DateTimeOffset start, DateTimeOffset end)
...
+            }
+
+            _logger.Debug("Adding peer {Peer} to routing table.", peer);
```
Comment (dahlia) [inline Libplanet/Net/Protocols/RoutingTable.cs:247]:
```suggestion
            _logger.Debug("Adding a peer {AddedPeer} to {Peer}'s routing table...", peer, this);
```

File: Libplanet.Tests/TestUtils.cs (hunk 1 above)
Comment (limebell) [reply Libplanet.Tests/TestUtils.cs:407]: Was used to count total tries for logging purpose, and it's not used anymore. I'll remove it thanks.

**Comment by riemannulus** [comment]
It seems to need a rebase

**Review by riemannulus — APPROVED** [review]

**Review by longfin — APPROVED** [review]

**Review by longfin — APPROVED** [review]

**Review by dahlia — APPROVED** [review]

**Review by moreal — APPROVED** [review]

File: Libplanet.Tests/Net/Protocols/RoutingTableTest.cs [hunk 4]
```diff
@@ -152,9 +165,223 @@ public void RemovePeer()
...
+
+            return (publicKey, publicKeys);
+        }
+
+        private string ByteArrayToString(byte[] bytes)
+        {
+            var str = BitConverter.ToString(bytes);
+            str = "0x" + str.Replace("-", ", 0x").ToLower();
+
+            return str + ",";
```
[the remark this answers is not in the document]
Comment (limebell) [reply Libplanet.Tests/Net/Protocols/RoutingTableTest.cs:384]: I've made `ByteArrayToString()` method to easily generate input of `PublicKey` constructor, such as:
```
                new PublicKey(
                    new byte[]
                    {
                        0x02, 0x4c, 0x11, 0x05, 0x7c, 0x7e, 0x18, 0xd6, 0x04, 0xd4, 0x5b, 0x01,
                        0xdb, 0xba, 0x6c, 0x03, 0xf0, 0x6c, 0x95, 0x45, 0xa9, 0x98, 0x50, 0xbc,
                        0x8e, 0x9b, 0x7c, 0x65, 0xbe, 0x41, 0x97, 0xce, 0x5c,
                    }),
```
The log here is
```
13:52:54:958691 - Found public key of bucket index 5 in 50 tries: "0x03, 0x1c, 0x1a, 0x0d, 0x8e, 0x0f, 0x25, 0x05, 0x40, 0xc5, 0xc7, 0x9d, 0x32, 0xc3, 0xfa, 0xef, 0xdb, 0xb4, 0x83, 0xc0, 0xa1, 0xb1, 0x22, 0x8c, 0x0b, 0x65, 0x32, 0xfe, 0xec, 0x57, 0x3f, 0x7c, 0x2d,"
```
So it's easy to make test cases.

File: Libplanet.Tests/Net/Protocols/RoutingTableTest.cs [hunk 5]
```diff
@@ -152,9 +165,223 @@ public void RemovePeer()
...
+                    }),
+            };
+
+            return (publicKey, publicKeys);
+        }
+
+        private string ByteArrayToString(byte[] bytes)
+        {
+            var str = BitConverter.ToString(bytes);
+            str = "0x" + str.Replace("-", ", 0x").ToLower();
```
[the remark this answers is not in the document]
Comment (moreal) [reply Libplanet.Tests/Net/Protocols/RoutingTableTest.cs:380]: It was my mistake. I'm sorry 😢

**Review by moreal — APPROVED** [review]

**Review by dahlia — APPROVED** [review]

**Review by longfin — APPROVED** [review]

**Review by riemannulus — APPROVED** [review]

**Pull request merged**